gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java

```diff
@@ -646,42 +646,48 @@ private TableId setProjectId(TableId table) {
   }
 
   private JobInfo setProjectId(JobInfo job) {
-    if (job instanceof CopyJobInfo) {
-      CopyJobInfo copyJob = (CopyJobInfo) job;
-      CopyJobInfo.Builder copyBuilder = copyJob.toBuilder();
-      copyBuilder.destinationTable(setProjectId(copyJob.destinationTable()));
-      copyBuilder.sourceTables(
-          Lists.transform(copyJob.sourceTables(), new Function<TableId, TableId>() {
-            @Override
-            public TableId apply(TableId tableId) {
-              return setProjectId(tableId);
-            }
-          }));
-      return copyBuilder.build();
-    }
-    if (job instanceof QueryJobInfo) {
-      QueryJobInfo queryJob = (QueryJobInfo) job;
-      QueryJobInfo.Builder queryBuilder = queryJob.toBuilder();
-      if (queryJob.destinationTable() != null) {
-        queryBuilder.destinationTable(setProjectId(queryJob.destinationTable()));
-      }
-      if (queryJob.defaultDataset() != null) {
-        queryBuilder.defaultDataset(setProjectId(queryJob.defaultDataset()));
-      }
-      return queryBuilder.build();
-    }
-    if (job instanceof ExtractJobInfo) {
-      ExtractJobInfo extractJob = (ExtractJobInfo) job;
-      ExtractJobInfo.Builder extractBuilder = extractJob.toBuilder();
-      extractBuilder.sourceTable(setProjectId(extractJob.sourceTable()));
-      return extractBuilder.build();
-    }
-    if (job instanceof LoadJobInfo) {
-      LoadJobInfo loadJob = (LoadJobInfo) job;
-      LoadJobInfo.Builder loadBuilder = loadJob.toBuilder();
-      return loadBuilder.configuration(setProjectId(loadJob.configuration())).build();
-    }
-    return job;
+    JobConfiguration configuration = job.configuration();
+    JobInfo.Builder jobBuilder = job.toBuilder();
+    switch (configuration.type()) {
```

Comment:
Did you considerhaving a package-scoped setProjectId in JobConfiguration that would return another updated JobConfiguration. I wonder if we can eliminate this switch+cast if we move each logic to its specific implementation?